File: apps/web/lib/api/companionWorkspace.ts
import { api, queryString } from "./client";

export interface CompanionWorkspaceReadModel {
  companion: { id: string; name: string; subtitle: string | null; current_mode: string; current_status: string | null; current_focus: string | null };
  identity: { display_name: string; identity_summary: string; core_traits: string[]; persona_summary: string; persona_lock_level: string; relationship_role: string; relationship_summary: string };
  boundary: { private_memory_default: string; shared_memory_default: string; cross_companion_read_policy: string; private_to_shared_review_required: boolean; shared_to_private_review_required: boolean; cross_companion_review_required: boolean };
  continuity: { conversation_id: string | null; current_topic: string | null; current_goal: string | null; current_phase: string | null; last_assistant_summary: string | null; suggested_next_steps: unknown[]; updated_at: string | null } | null;
  relationship: Record<string, string | number | null> | null;
  recent_private_memories: Array<{ id: string; type: string; summary: string; updated_at: string | null }>;
  presence_preview: Array<{ id: string; type: string; title: string; message: string | null; priority: number; recommended_surface: string; expires_at: string | null }>;
  review_counts: Record<string, number>;
  governance: { hard_stop_active: boolean; hard_stop_scope: string | null; revoked_channels: number; active_channels: number };
  channels?: Array<{ id: string; status: string; scope: string; outbound_policy: string; memory_review_required: boolean }>;
  channel_presence?: Array<{ status: string; mode: string; muted: boolean; checkin_enabled: boolean; quiet_hours: boolean }>;
  voice?: { profile_status: string | null; profile_name: string | null; session_status: string | null; transcript_retention: string | null; memory_write_policy: string | null; real_audio_enabled: boolean };
}

export interface ChronicleReadModel {
  companion_id: string;
  items: Array<{ id: string; companion_id: string; kind: string; occurred_at: string; title: string; summary: string; source_id: string | null; review_status: string | null; trace_id: string | null }>;
  summaries: Array<{ id: string; version: number; status: string; title: string; summary: string; highlights: string[]; period_start: string; period_end: string; source_event_refs: string[]; invalidation_reason: string | null; created_at: string }>;
  total: number;
  limit: number;
  offset: number;
}

export interface ReviewInboxReadModel {
  companion_id: string;
  items: Array<{ id: string; companion_id: string; kind: string; created_at: string; title: string; summary: string; status: string; risk_level: string | null; source_id: string | null }>;
  counts: Record<string, number>;
  total: number;
  limit: number;
  offset: number;
}

export const companionWorkspaceApi = {
  workspace: (companionId: string) => api.get<CompanionWorkspaceReadModel>(`/companions/${companionId}/workspace`),
  chronicle: (companionId: string, limit = 40, offset = 0) => api.get<ChronicleReadModel>(`/companions/${companionId}/chronicle${queryString({ limit, offset })}`),
  refreshChronicleSummary: (companionId: string, correctionNote?: string) => api.post(`/companions/${companionId}/chronicle/summaries/refresh`, { correction_note: correctionNote || null }),
  invalidateChronicleSummary: (companionId: string, summaryId: string, reason: string) => api.post(`/companions/${companionId}/chronicle/summaries/${summaryId}/invalidate`, { reason }),
  reviewInbox: (companionId: string, limit = 50, offset = 0, kind?: string) => api.get<ReviewInboxReadModel>(`/companions/${companionId}/review-inbox${queryString({ limit, offset, kind })}`),
  reviewInboxPage: async (companionId: string, limit: number, offset: number, kind?: string) => {
    const page = await companionWorkspaceApi.reviewInbox(companionId, limit, offset, kind);
    if (!kind || (page.total === page.counts[kind] && page.items.every((item) => item.kind === kind))) return page;

    const allItems: ReviewInboxReadModel["items"] = [];
    const total = page.counts.total ?? page.total;
    for (let batchOffset = 0; batchOffset < total; batchOffset += 100) {
      const batch = await companionWorkspaceApi.reviewInbox(companionId, 100, batchOffset);
      allItems.push(...batch.items);
    }
    const filtered = allItems.filter((item) => item.kind === kind);
    return { ...page, items: filtered.slice(offset, offset + limit), total: filtered.length, limit, offset };
  },
  decidePersonaGrowth: (companionId: string, candidateId: string, decision: "approved" | "rejected") => api.post(`/companions/${companionId}/persona-growth-candidates/${candidateId}/decision`, { decision }),
};

File: services/agent-api/app/schemas/companion_workspace.py
"""Read-only contracts for the Companion-first frontend."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class WorkspaceCompanion(BaseModel):
    id: str
    name: str
    subtitle: str | None = None
    current_mode: str
    current_status: str | None = None
    current_focus: str | None = None


class WorkspaceIdentity(BaseModel):
    display_name: str
    identity_summary: str
    core_traits: list[str] = Field(default_factory=list)
    persona_summary: str
    persona_lock_level: str
    relationship_role: str
    relationship_summary: str


class WorkspaceBoundary(BaseModel):
    private_memory_default: str
    shared_memory_default: str
    cross_companion_read_policy: str
    private_to_shared_review_required: bool
    shared_to_private_review_required: bool
    cross_companion_review_required: bool


class WorkspaceContinuity(BaseModel):
    conversation_id: str | None = None
    current_topic: str | None = None
    current_goal: str | None = None
    current_phase: str | None = None
    last_assistant_summary: str | None = None
    suggested_next_steps: list[Any] = Field(default_factory=list)
    updated_at: datetime | None = None


class WorkspaceMemoryPreview(BaseModel):
    id: str
    type: str
    summary: str
    updated_at: datetime | None = None


class WorkspacePresencePreview(BaseModel):
    id: str
    type: str
    title: str
    message: str | None = None
    priority: float
    recommended_surface: str
    expires_at: datetime | None = None


class CompanionWorkspaceReadModel(BaseModel):
    companion: WorkspaceCompanion
    identity: WorkspaceIdentity
    boundary: WorkspaceBoundary
    continuity: WorkspaceContinuity | None = None
    relationship: dict[str, Any] | None = None
    recent_private_memories: list[WorkspaceMemoryPreview] = Field(default_factory=list)
    presence_preview: list[WorkspacePresencePreview] = Field(default_factory=list)
    review_counts: dict[str, int] = Field(default_factory=dict)


class ChronicleItem(BaseModel):
    id: str
    companion_id: str
    kind: str
    occurred_at: datetime
    title: str
    summary: str
    source_id: str | None = None
    review_status: str | None = None
    trace_id: str | None = None


class CompanionChronicleReadModel(BaseModel):
    companion_id: str
    items: list[ChronicleItem]
    total: int
    limit: int
    offset: int


ReviewKind = Literal[
    "memory", "growth", "persona_growth", "private_to_shared",
    "shared_to_private", "cross_companion", "channel", "realtime_shared",
]


class ReviewInboxItem(BaseModel):
    id: str
    companion_id: str
    kind: ReviewKind
    created_at: datetime
    title: str
    summary: str
    status: str
    risk_level: str | None = None
    source_id: str | None = None


class CompanionReviewInboxReadModel(BaseModel):
    companion_id: str
    items: list[ReviewInboxItem]
    counts: dict[str, int]
    total: int
    limit: int
    offset: int
